src/server.ts

```diff
@@ -178,56 +222,66 @@ class Server {
           // find subscription id. Call unsubscribe.
           // TODO untested. catch errors, etc.
           if (typeof connectionSubscriptions[subId] !== 'undefined') {
-            this.subscriptionManager.unsubscribe(connectionSubscriptions[subId]);
+            this.unsubscribe(connection, connectionSubscriptions[subId]);
             delete connectionSubscriptions[subId];
           }
           break;
 
         default:
           this.sendSubscriptionFail(connection, subId, {
             errors: [{
-              message: 'Invalid message type. Message type must be `subscription_start` or `subscription_end`.'
-            }]
+              message: 'Invalid message type!.',
+            }],
           });
       }
     };
   }
 
-  private sendSubscriptionData(connection: Connection, subId: string, payload: any): void {
+  private sendSubscriptionData(connection: WebSocket, subId: string, payload: any): void {
     let message = {
       type: SUBSCRIPTION_DATA,
       id: subId,
       payload,
     };
 
-    connection.sendUTF(JSON.stringify(message));
+    connection.send(JSON.stringify(message));
   }
 
-  private sendSubscriptionFail(connection: Connection, subId: string, payload: any): void {
+  private sendSubscriptionFail(connection: WebSocket, subId: string, payload: any): void {
     let message = {
       type: SUBSCRIPTION_FAIL,
       id: subId,
       payload,
     };
 
-    connection.sendUTF(JSON.stringify(message));
+    connection.send(JSON.stringify(message));
   }
 
-  private sendSubscriptionSuccess(connection: Connection, subId: string): void {
+  private sendSubscriptionSuccess(connection: WebSocket, subId: string): void {
     let message = {
       type: SUBSCRIPTION_SUCCESS,
       id: subId,
     };
 
-    connection.sendUTF(JSON.stringify(message));
+    connection.send(JSON.stringify(message));
+  }
+
+  private sendInitResult(connection: WebSocket, result: any): void {
+    connection.send(JSON.stringify(result), () => {
+      if (result.type === INIT_FAIL) {
+        connection.close();
```

Comment:
Are you sure that this actually works?  Does the close packet actually get sent out over the network or does calling terminate synchronously close the socket too fast (I think it does).  You can test this with wireshark or tcpdump or something...